bin/varnishd/cache/cache_req_body.c

```diff
@@ -259,17 +259,12 @@ VRB_Ignore(struct req *req)
 void
 VRB_Free(struct req *req)
 {
-	int r;
-
 	CHECK_OBJ_NOTNULL(req, REQ_MAGIC);
 
 	if (req->body_oc == NULL)
 		return;
 
-	r = HSH_DerefObjCore(req->wrk, &req->body_oc, 0);
-
-	// a busyobj may have gained a reference
-	assert (r == 0 || r == 1);
```

Comment:
alternative suggestion:
```diff
diff --git a/bin/varnishd/cache/cache_req_body.c b/bin/varnishd/cache/cache_req_body.c
index a3a771163..4cb01e815 100644
--- a/bin/varnishd/cache/cache_req_body.c
+++ b/bin/varnishd/cache/cache_req_body.c
@@ -268,8 +268,8 @@ VRB_Free(struct req *req)
 
        r = HSH_DerefObjCore(req->wrk, &req->body_oc, 0);
 
-       // a busyobj may have gained a reference
-       assert (r == 0 || r == 1);
+       // each busyobj may have gained a reference
+       assert (r <= req->restarts + 1);
 }
 
 /*----------------------------------------------------------------------

```
?

Reply:
Shouldn't we make `HSH_DerefObjCore()` return `unsigned` in addition? It doesn't seem to ever return a negative value.

Reply:
yes

Reply:
I feels a bit odd to assert that we are less than the number of restarts. Is there a particular refcount scenario we are guarding against? Most usages of `HSH_DerefObjCore()` do not bother trying to do this. Im ok leaving this as is.

Reply:
This is to guard against reference leaks.

Reply:
We can still leak and pass thru this assert. We could in theory pull extra valid references and then incorrectly trigger this assert (for example writing the req.body somewhere else). So that's why I feel this doesn't really accomplish much.

Reply:
That's basically to maintain the role of the previous assert that forgot to take restarts into account.

Reply:
Right, this is what I'm saying, the previous assert doesn't accomplish anything. It doesn't protect from actual leaks and only serves as a hidden tripwire.

Reply:
Both asserts are equivalent for the common `req->restarts == 0` case .
In that case, if we were to accidentally get a second reference, we would trigger.
So the assert *does* accomplish something and it *does* protect from actual leaks.
Also, even for the case which you have spotted, it made us think more and understand the code better.

Reply:
> In that case, if we were to accidentally get a second reference, we would trigger.

No, because that reference could be properly returned or taken after `VRB_Free()`. Or it could be taken, lost, and then it falls under that extra `+ 1` in the calculation. Again, `refcnt` cannot be safely used to tell us anything in terms of the number of references taken or potential leaks.

Im ok putting a comment in place of the assert describing all the paths a reference could be taken. What do you think?

Reply:
this is ridiculous. Unfortunately. Again.